File: runners/mastra/runner/tasks/task05.mjs
/**
 * Task 5 — Mastra context-tax demo (the "context scrubbing" competitor arm), the
 * TypeScript sibling of runners/pydantic_ai/runner/tasks/task05.py.
 *
 * Replays the fixed 10-turn conversation with Mastra's idiomatic multi-turn
 * pattern: each turn calls agent.generate(messages, ...) with the FULL running
 * transcript, and the assistant + tool messages returned are appended back for the
 * next turn. That is the framework's default memory — the whole history, including
 * the ~32 KB base64 chart tool-return, is re-sent every turn and never trimmed.
 * This is the competitor baseline that Colmena's binary scrubber eliminates.
 *
 * Seeding: the report is planted as a pre-turn-0 user/assistant exchange (no LLM
 * call), matching how the other runners seed a static report + acknowledgement.
 *
 * Chart payload transport workaround: LiteLLM's Gemini translator auto-promotes any
 * tool/message text starting with "data:image/" into a Gemini image part (rejected
 * for a synthetic PNG). We prefix the payload with "[chart_data_uri]: " so it does
 * NOT start with "data:" — the full ~32 KB is still present as text, so the
 * context-tax growth is faithfully measured. (Every other runner does this.)
 *
 * Token accounting is via proxy spans bucketed by extras.turn_boundaries; the
 * returned usage is all zeros by contract.
 */
import { Agent } from '@mastra/core/agent';
import { createTool } from '@mastra/core/tools';
import { z } from 'zod';

import { S05 } from '../fixtures.mjs';

const nowIso = () => new Date().toISOString();

export async function run(taskDef, model, args) {
  const chartTool = createTool({
    id: S05.chart_tool_name, // "generate_chart"
    description: S05.chart_tool_description,
    inputSchema: z.object({ description: z.string() }),
    outputSchema: z.object({ chart: z.string() }),
    // Fixed payload regardless of input; prefixed so it does not start with data:.
    execute: async () => ({ chart: `[chart_data_uri]: ${S05.chart_data_uri}` }),
  });

  const agent = new Agent({
    name: 'ReportAnalyst',
    instructions: S05.system_message,
    model,
    tools: { [S05.chart_tool_name]: chartTool },
  });

  // Seed the report as a pre-turn-0 exchange (no LLM call), mirroring the other
  // runners' static system + report + acknowledgement seed.
  const messages = [
    { role: 'user', content: `Here is the report for this conversation:\n\n${S05.report_text}` },
    { role: 'assistant', content: 'Understood. I have the report and will answer your questions.' },
  ];

  const answers = [];
  const turnBoundaries = [nowIso()]; // boundary BEFORE turn 0

  for (let i = 0; i < S05.turns.length; i += 1) {
    const turn = S05.turns[i];
    messages.push({ role: 'user', content: turn.message });
    try {
      const res = await agent.generate(messages, { modelSettings: { temperature: 0.0 } });
      const text = String((res && res.text) || '');
      answers.push(text);
      // Append the full assistant/tool step messages so the ~32 KB chart return
      // stays in the re-sent transcript for every later turn (the context tax).
      const resp = res && res.response ? await res.response : null;
      const stepMsgs = (resp && resp.messages) || null;
      if (Array.isArray(stepMsgs) && stepMsgs.length) {
        for (const m of stepMsgs) messages.push(m);
      } else {
        messages.push({ role: 'assistant', content: text });
      }
    } catch (e) {
      answers.push(`[ERROR turn ${i}: ${e && e.message ? e.message : e}]`);
      messages.push({ role: 'assistant', content: '' });
    } finally {
      turnBoundaries.push(nowIso()); // boundary AFTER this turn
    }
  }

  return {
    answer: answers,
    usage: { input: 0, output: 0, cached: 0, tool_calls: 0 },
    extras: {
      turn_boundaries: turnBoundaries,
      turn_types: S05.turns.map((t) => t.type),
    },
  };
}

File: runners/pydantic_ai/runner/tasks/task05.py
"""Task 5 — Pydantic AI context-tax demo (the "context scrubbing" competitor arm).

Replays the fixed 10-turn conversation (``bench_common.TURNS``) with Pydantic AI's
idiomatic multi-turn pattern: each turn calls ``agent.run_sync(msg,
message_history=...)`` and feeds the FULL returned message list back into the next
turn. That is the framework's default memory — the whole history, including the
~32 KB base64 chart tool-return, is re-sent every turn and never trimmed. This is
the competitor baseline that Colmena's binary scrubber eliminates.

Seeding: the report is planted as a pre-turn-0 exchange built directly from
``ModelRequest``/``ModelResponse`` (no LLM call), matching how the LangChain runner
seeds a static ``HumanMessage``/``AIMessage`` pair. The system prompt is set on the
``Agent``.

Chart payload transport workaround: LiteLLM's Gemini translator auto-promotes any
tool/message text starting with ``data:image/`` into a Gemini image part (which it
then rejects for a synthetic PNG). We prefix the payload with ``[chart_data_uri]: ``
so it does NOT start with ``data:`` — the full ~32 KB is still present as text, so
the context-tax growth is faithfully measured. (Every other runner does this.)

Token accounting is via proxy spans bucketed by ``extras.turn_boundaries``; the
returned ``usage`` is all zeros by contract.
"""
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from pydantic_ai import Agent
from pydantic_ai.messages import ModelRequest, ModelResponse, TextPart, UserPromptPart

from bench_common import (
    CHART_TOOL_DESCRIPTION,
    REPORT_TEXT,
    SYSTEM_MESSAGE,
    TURNS,
    RunnerArgs,
    generate_chart as _generate_chart,
)


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


def run(
    task_def: dict[str, Any], model: Any, args: RunnerArgs
) -> tuple[list[str], dict[str, int], dict[str, Any]]:
    agent = Agent(
        model,
        system_prompt=SYSTEM_MESSAGE,
        model_settings={"temperature": 0.0},
    )

    # The tool name doubles as CHART_TOOL_NAME ("generate_chart"); its docstring is
    # CHART_TOOL_DESCRIPTION so the schema the model sees matches the other runners.
    @agent.tool_plain
    def generate_chart(description: str) -> str:
        return f"[chart_data_uri]: {_generate_chart(description)}"

    generate_chart.__doc__ = CHART_TOOL_DESCRIPTION

    # Seed the report as a pre-turn-0 exchange (no LLM call), mirroring the other
    # runners' static system + report + acknowledgement seed.
    history: list[Any] = [
        ModelRequest(parts=[
            UserPromptPart(content=f"Here is the report for this conversation:\n\n{REPORT_TEXT}")
        ]),
        ModelResponse(parts=[
            TextPart(content="Understood. I have the report and will answer your questions.")
        ]),
    ]

    answers: list[str] = []
    turn_boundaries: list[str] = [_now_iso()]  # boundary BEFORE turn 0

    for i, turn in enumerate(TURNS):
        try:
            result = agent.run_sync(turn["message"], message_history=history)
            history = result.all_messages()  # full verbatim history -> next turn
            answers.append(str(getattr(result, "output", "") or ""))
        except Exception as e:  # noqa: BLE001 — one bad turn must not sink the run
            answers.append(f"[ERROR turn {i}: {type(e).__name__}: {e}]")
        finally:
            turn_boundaries.append(_now_iso())  # boundary AFTER this turn

    usage = {"input": 0, "output": 0, "cached": 0, "tool_calls": 0}
    extras = {
        "turn_boundaries": turn_boundaries,
        "turn_types": [t["type"] for t in TURNS],
    }
    return answers, usage, extras
